src/minimization.cpp

```diff
@@ -0,0 +1,211 @@
+        edges.emplace_back(successorNode->number, j);
+      }
+    }
+    std::sort(edges.begin(), edges.end());
+    for (int j = 0; j < edges.size() - 1; ++j) {
+      for (int k = 0; k < edges.size() - 1; ++k) {
+        if (edges[k].second > edges[k + 1].second) {
+          std::swap(edges[k], edges[k + 1]);
+          crossNum += 1;
+        }
+      }
+    }
+  }
+  return crossNum;
+}
+
+void sortPorts(Net& net, std::vector<std::vector<TreeNode::nodeId>>& nodesByLayer) {
+  for (int i = 1; i < nodesByLayer.size(); ++i) {
+    for (int j = 0; j < nodesByLayer[i].size(); ++j) {
+      TreeNode *node = net.getNode(nodesByLayer[i][j]);
+      std::vector<std::pair<float, TreeNode::nodeId>> bPred, bSucc;
+      for (TreeNode::nodeId &predecessor : node->pred) {
+        TreeNode *predecessorNode = net.getNode(predecessor);
+        float bValue = forwardRankDefinition(*predecessorNode, predecessorNode->number,
+          getPortIndex(node->id, *predecessorNode)) /
+          float(node->succ.size() + node->pred.size());
+        bPred.push_back({ bValue, predecessorNode->id });
+      }
+      for (TreeNode::nodeId &successor : node->succ) {
+        TreeNode *successorNode = net.getNode(successor);
+        float bValue = backwardRankDefinition(*successorNode, successorNode->number,
+          getPortIndex(node->id, *successorNode)) /
+          float(node->succ.size() + node->pred.size());
+        bSucc.push_back({ bValue, successorNode->id });
+      }
+      std::sort(bPred.begin(), bPred.end());
+      std::sort(bSucc.begin(), bSucc.end());
+
+      for (int k = 0; k < bPred.size(); ++k) {
+        node->pred[k] = bPred[k].second;
+      }
+      for (int k = 0; k < bSucc.size(); ++k) {
+        node->succ[k] = bSucc[k].second;
+      }
+    }
+  }
+}
+
+void barycentricValueDefinition(Net &net, TreeNode *node, int increment) {
+  int connectionsToAdjacentLayer = 0;
+  float rank = 0;
+  for (int k = 0; k < node->pred.size(); ++k) {
+    if (net.getNode(node->pred[k])->layer + increment != node->layer)
+      continue;
+    connectionsToAdjacentLayer += 1;
+    int index = net.getNode(node->pred[k])->number;
+    int portIndex = getPortIndex(node->id, * net.getNode(node->pred[k]));
+    if (increment == 1) {
+      rank += forwardRankDefinition( *net.getNode(node->pred[k]), index, portIndex);
+    } else {
+      rank += backwardRankDefinition( *net.getNode(node->pred[k]), index, portIndex);
+    }
+  }
+  for (int k = 0; k < node->succ.size(); ++k) {
+    if (net.getNode(node->succ[k])->layer + increment != node->layer)
+      continue;
+    connectionsToAdjacentLayer += 1;
+    int index = net.getNode(node->succ[k])->number;
+    int portIndex = getPortIndex(node->id, *net.getNode(node->succ[k]));
+    if (increment == 1) {
+      rank += forwardRankDefinition( *net.getNode(node->pred[k]), index, portIndex);
+    } else {
+      rank += backwardRankDefinition( *net.getNode(node->succ[k]), index, portIndex);
+    }
+  }
```

Comment:
Code duplication. Actually, it is suspicious that the first loop uses `xxxRankDefinition` on a predecessor, but the second loop uses it on a predecessor *or* a successor. Is it a bug?